### PR TITLE
Fix new dialog creation, if no input var in terraform template and spec added

### PR DESCRIPTION
**Pull request by putmanoj**
* Fix new dialog creation, if no input var in terraform template 
* Add spec's terraform template dynamic dialog creation to cover various scenarios.

### CODE REVIEW

File: app/models/dialog/terraform_template_service_dialog.rb [hunk 1]
```diff
@@ -9,13 +9,20 @@ def create_dialog(label, terraform_template, extra_vars)
...
         end
         if extra_vars.present?
           add_variables_group(tab, position, extra_vars)
+          position += 1
+        end
+
+        if position == 0
+          # if not template input vars & no extra_vars,
+          # then add single variable text box as place-holder
+          add_variables_group(tab, position, {:name=>{:default=>label}})
```
Comment (agrare) [inline app/models/dialog/terraform_template_service_dialog.rb:23]:
```suggestion
          add_variables_group(tab, position, {:name => {:default => label}})
```

**Comment by agrare** [comment]
With the relatively complex number of scenarios now and the fact an exception already slipped through I think it'd be a good idea to write some spec tests covering this

**Comment by putmanoj** [comment]
> With the relatively complex number of scenarios now and the fact an exception already slipped through I think it'd be a good idea to write some spec tests covering this

Yes, will add ..

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      end
+    end
+  end
+
+  def assert_variables_tab(dialog, group_size: 1)
+    tabs = dialog.dialog_tabs
+    expect(tabs.size).to eq(1)
+
+    tab0 = tabs[0]
+    assert_tab_attributes(tab0)
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:142]: Minor formatting concern
```suggestion
    assert_tab_attributes(tabs.first)
```

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+      let(:extra_vars) do
+        {}
+      end
+
+      it_behaves_like "create_dialog with place-holder variable argument"
+    end
+
+    context "when nil terraform template & nil extra vars" do
+      let(:dialog_label) { "mydialog3" }
+
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:99]:
```suggestion
```

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      let(:terraform_template) { nil }
+
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:101]:
```suggestion
```

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+      let(:input_vars) do
+        require 'json'
+        payload = JSON.parse(payload_with_three_input_vars)
+        payload['input_vars']
+      end
+
+      it "creates multiple dialogs" do
+        dialog = subject.create_dialog(dialog_label, terraform_template_with_input_vars, extra_vars)
+        expect(dialog).to have_attributes(:label => dialog_label, :buttons => "submit,cancel")
+ \
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:127]:
```suggestion

```

**Comment by putmanoj** [comment]
> > With the relatively complex number of scenarios now and the fact an exception already slipped through I think it'd be a good idea to write some spec tests covering this
> 
> Yes, will add ..

Spec added

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    tabs = dialog.dialog_tabs
+    expect(tabs.size).to eq(1)
+
+    tab0 = tabs[0]
+    assert_tab_attributes(tab0)
+
+    groups = tab0.dialog_groups
+    expect(groups.size).to eq(group_size)
+
+    groups[group_size > 1 ? 1 : 0]
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:147]: Was this supposed to be in an expect ?

Comment (agrare) [reply spec/models/dialog/terraform_template_service_dialog_spec.rb:147]: Oh interesting you're returning it from this method okay

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+  end
+
+  def assert_terraform_template_variables_tab(dialog, group_size: 1)
+    tabs = dialog.dialog_tabs
+    expect(tabs.size).to eq(1)
+
+    tab0 = tabs[0]
+    assert_tab_attributes(tab0)
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:155]:
```suggestion
    assert_tab_attributes(tabs.first)
```

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+    groups = tab0.dialog_groups
+    expect(groups.size).to eq(group_size)
+
+    group0 = groups[0]
+
+    expect(group0).to have_attributes(:label => "Terraform Template Variables", :display => "edit")
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:162]:
```suggestion

    expect(group.first).to have_attributes(:label => "Terraform Template Variables", :display => "edit")
```

File: spec/models/dialog/terraform_template_service_dialog_spec.rb [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+    group0
+  end
+
+  def assert_tab_attributes(tab)
+    expect(tab).to have_attributes(:label => "Basic Information", :display => "edit")
+  end
+
+  def assert_field(field, clss, attributes)
+    expect(field).to be_kind_of clss
```
Comment (agrare) [inline spec/models/dialog/terraform_template_service_dialog_spec.rb:172]: Really minor but we tend to use `klass` for a class descriptor

**Comment by agrare** [comment]
All comments minor, https://github.com/ManageIQ/manageiq-providers-embedded_terraform/pull/77/files#r1807080713 is the only one I really don't want to merge as-is

**Review by agrare — APPROVED** [review]

**Comment by Fryguy** [comment]
Backported to `radjabov` in commit dbf6da49ba8c4e019742bbc3cab5596279f406e5.

```text
commit dbf6da49ba8c4e019742bbc3cab5596279f406e5
Author: Adam Grare <adam@grare.com>
Date:   Tue Oct 22 09:04:44 2024 -0400

    Merge pull request #77 from putmanoj/fix-dialog-if-no-input-var
    
    Fix new dialog creation, if no input var in terraform template and spec added
    
    (cherry picked from commit 2f44806b4b64cc56a68fb2e6af6cbe23149e798e)
```

**Pull request merged**